--- a/proc.c
+++ b/proc.c
@@ -37,7 +37,9 @@
 	__header(content_length, content-length) \
 	__header(x_http_method_override, x-http-method-override) \
 	__header(http_auth_user, http-auth-user) \
-	__header(http_auth_pass, http-auth-pass)
+	__header(http_auth_pass, http-auth-pass) \
+	__header(x_real_ip, x-real-ip) \
+	__header(x_real_port, x-real-port)
 
 #undef __header
 #define __header __enum_header
@@ -163,6 +165,20 @@ struct env_var *uh_get_process_vars(stru
 	snprintf(remote_port, sizeof(remote_port), "%d", cl->peer_addr.port);
 
 	blobmsg_parse(hdr_policy, __HDR_MAX, tb, blob_data(data), blob_len(data));
+
+	if (tb[HDR_x_real_ip]) {
+		char* realIp = blobmsg_data(tb[HDR_x_real_ip]);
+		if (realIp && strlen(realIp)) {
+			strncpy(remote_addr, realIp, sizeof(remote_addr) - 1);
+		}
+	}
+	if (tb[HDR_x_real_port]) {
+		char* realPort = blobmsg_data(tb[HDR_x_real_port]);
+		if (realPort && strlen(realPort)) {
+			strncpy(remote_port, realPort, sizeof(remote_port) - 1);
+		}
+	}
+
 	for (i = 0; i < ARRAY_SIZE(proc_header_env); i++) {
 		struct blob_attr *cur;
 
