diff --git a/package/network/config/netifd/files/etc/init.d/network b/package/network/config/netifd/files/etc/init.d/network
index dc208c4..de366fa 100755
--- a/package/network/config/netifd/files/etc/init.d/network
+++ b/package/network/config/netifd/files/etc/init.d/network
@@ -40,6 +40,14 @@ stop_service() {
 	sleep 1
 }
 
+
+
+
+service_running() {
+        ubus -t 120 wait_for network.interface
+        /sbin/wifi reload_legacy
+}
+
 validate_atm_bridge_section()
 {
 	uci_validate_section network "atm-bridge" "${1}" \
@@ -139,6 +147,8 @@ service_triggers()
 }
 
 shutdown() {
-	ifdown -a
-	sleep 1
+        /sbin/wifi down
+        ifdown -a
+        sleep 1
 }
+
