diff -urN ori/scripts/Makefile new/scripts/Makefile
--- ori/scripts/Makefile	2021-09-16 17:06:22.698558000 +0800
+++ new/scripts/Makefile	2021-09-16 17:06:53.982483000 +0800
@@ -10,7 +10,7 @@
 
 HOST_EXTRACFLAGS += -I$(srctree)/tools/include
 
-CRYPTO_LIBS = $(shell pkg-config --libs libcrypto 2> /dev/null || echo -lcrypto)
+CRYPTO_LIBS = $(shell pkg-config --libs libcrypto 2> /dev/null || echo -lcrypto) -pthread
 CRYPTO_CFLAGS = $(shell pkg-config --cflags libcrypto 2> /dev/null)
 
 hostprogs-$(CONFIG_BUILD_BIN2C)  += bin2c
