From 1bea49e81bc578e6cbbdde5e0d4276f05fb69866 Mon Sep 17 00:00:00 2001
From: caojingzhe <caojingzhe@twsh.com>
Date: Tue, 24 May 2022 16:24:35 +0800
Subject: [PATCH] 2

---
 .../api/wext.h                            | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/api/wext.h b/api/wext.h
index 6b5838e1..40eefa72 100644
--- a/api/wext.h
+++ b/api/wext.h
@@ -543,7 +543,7 @@
 #define IW_SCAN_TYPE_ACTIVE 0
 #define IW_SCAN_TYPE_PASSIVE 1
 /* Maximum size of returned data */
-#define IW_SCAN_MAX_DATA	4096	/* In bytes */
+#define IW_SCAN_MAX_DATA	(4096*8)	/* In bytes */
 
 /* Max number of char in custom event - use multiple of them if needed */
 #define IW_CUSTOM_MAX		256	/* In bytes */
-- 
2.13.1

