--- a/mt_wifi/embedded/common/cmm_info.c	2022-04-28 10:57:41.749887972 +0800
+++ b/mt_wifi/embedded/common/cmm_info.c	2022-04-28 14:32:57.239764822 +0800
@@ -6245,6 +6245,15 @@
 		MTWF_DBG(NULL, DBG_CAT_AP, DBG_SUBCAT_ALL, DBG_LVL_ERROR, "Snprintf failed!\n");
 		return;
 	}
+
+	/* SSID Length */
+	LeftBufSize = MsgLen - strlen(msg);
+	ret = snprintf(msg + strlen(msg), LeftBufSize, " %-8d", pBss->SsidLen);
+	if (os_snprintf_error(LeftBufSize, ret)) {
+		MTWF_DBG(NULL, DBG_CAT_AP, DBG_SUBCAT_ALL, DBG_LVL_ERROR, "Snprintf failed!\n");
+		return;
+	}
+
 	/*SSID*/
 	NdisZeroMemory(Ssid, (MAX_LEN_OF_SSID + 1));
 
@@ -6438,13 +6447,6 @@
 		}
 	}
 
-	/* SSID Length */
-	LeftBufSize = MsgLen - strlen(msg);
-	ret = snprintf(msg + strlen(msg), LeftBufSize, " %-8d", pBss->SsidLen);
-	if (os_snprintf_error(LeftBufSize, ret)) {
-		MTWF_DBG(NULL, DBG_CAT_AP, DBG_SUBCAT_ALL, DBG_LVL_ERROR, "Snprintf failed!\n");
-		return;
-	}
 #ifdef CUSTOMER_VENDOR_IE_SUPPORT
 #ifdef CUSTOMER_MAXBITRATE_SUPPORT
 	sprintf(msg + strlen(msg), " %-8lld", pBss->CustomerBssEntry.max_bit_rate);
@@ -6605,7 +6607,7 @@
 #ifdef CUSTOMER_MAXBITRATE_SUPPORT
 	LeftBufSize = TotalLen - strlen(msg);
 	Status = snprintf(msg + strlen(msg), LeftBufSize, "%-4s%-4s%-33s%-20s%-23s%-9s%-11s%-7s%-3s%-8s%-16s\n",
-				"No", "Ch", "SSID", "BSSID", "Security", "Siganl(%)", "W-Mode", " ExtCH", " NT", " SSID_Len",
+				"No", "Ch", " SSID_Len", "SSID", "BSSID", "Security", "Siganl(%)", "W-Mode", " ExtCH", " NT",
 				"MaxBitRate");
 	if (os_snprintf_error(LeftBufSize, Status)) {
 		MTWF_DBG(pAdapter, DBG_CAT_CFG, DBG_SUBCAT_ALL, DBG_LVL_ERROR, "Snprintf failed!\n");
@@ -6615,10 +6617,10 @@
 	LeftBufSize = TotalLen - strlen(msg);
 #ifdef CCAPI_API_SUPPORT
 	Status = snprintf(msg + strlen(msg), LeftBufSize, "%-4s%-4s%-33s%-20s%-23s%-8s%-9s%-11s%-7s%-3s%-8s\n",
-				"No", "Ch", "SSID", "BSSID", "Security", "Rssi", "Siganl(%)", "W-Mode", " ExtCH", " NT", " SSID_Len");
+				"No", "Ch", " SSID_Len", "SSID", "BSSID", "Security", "Rssi", "Siganl(%)", "W-Mode", " ExtCH", " NT");
 #else
 	Status = snprintf(msg + strlen(msg), LeftBufSize, "%-4s%-4s%-33s%-20s%-23s%-9s%-11s%-7s%-3s%-8s\n",
-				"No", "Ch", "SSID", "BSSID", "Security", "Siganl(%)", "W-Mode", " ExtCH", " NT", " SSID_Len");
+				"No", "Ch", " SSID_Len", "SSID", "BSSID", "Security", "Siganl(%)", "W-Mode", " ExtCH", " NT");
 #endif
 	if (os_snprintf_error(LeftBufSize, Status)) {
 		MTWF_DBG(pAdapter, DBG_CAT_CFG, DBG_SUBCAT_ALL, DBG_LVL_ERROR, "Snprintf failed!\n");
